File: api_trial.py
# coding=utf-8
from __future__ import print_function
import os
import sys
import sqlite3
import json
import re
from indic_transliteration import sanscript
import xml.etree.ElementTree as ET
from flask import Flask, jsonify
from flask_restplus import Api, Resource, reqparse
from flask_cors import CORS
try:
	from HTMLParser import HTMLParser
except ImportError:
	from html.parser import HTMLParser

# Start Flask app.
app = Flask(__name__)
app.config['JSON_AS_ASCII'] = False
# Show request duration.
app.config.SWAGGER_UI_REQUEST_DURATION = True
CORS(app)
apiversion = 'v0.0.1'
api = Api(app, version=apiversion, title=u'Cologne Sanskrit-lexicon API', description='Provides APIs to Cologne Sanskrit lexica.')


# List of 34 open domain cologne dictionaries.
cologne_dicts = ['acc', 'ae', 'ap90', 'ben', 'bhs', 'bop', 'bor', 'bur', 'cae', 'ccs', 'gra', 'gst', 'ieg', 'inm', 'krm', 'mci', 'md', 'mw', 'mw72', 'mwe', 'pe', 'pgn', 'pui', 'pw', 'pwg', 'sch', 'shs', 'skd', 'snp', 'stc', 'vcp', 'vei', 'wil', 'yat']



def find_sqlite(dictionary):
	"""Return path to the sqlite file based on cologne server or local."""

	path = os.path.abspath(__file__)
	# If on Cologne server,
	if path.startswith('/nfs/'):
		intermediate = os.path.join(dict.upper() + 'Scan', '2020', 'web', 'sqlite', dictionary + '.sqlite')
	else:
		intermediate = dictionary
	sqlitepath = os.path.join('..', intermediate, 'web', 'sqlite', dictionary + '.sqlite')
	return sqlitepath


def convert_sanskrit(text, inTran, outTran):
	"""Return transliterated adjusted text."""

	text1 = ''
	counter = 0
	# Remove '<srs/>'
	text = text.replace('<srs/>', '')
	# Change the s tag to span.
	for i in re.split('<s>([^<]*)</s>', text):
		if counter % 2 == 0:
			text1 += i
		else:
			text1 += '<span class="s">' + sanscript.transliterate(i, 'slp1', outTran) + '</span>'
		counter += 1
	# PE nesting of LB tag
	text1 = text1.replace('<div n="1"/>', 'emsp;<div n="1"></div>')
	text1 = text1.replace('<div n="2"/>', 'emsp;emsp;<div n="2"></div>')
	text1 = text1.replace('<div n="3"/>', 'emsp;emsp;emsp;<div n="3"></div>')
	text1 = text1.replace('<div n="4"/>', 'emsp;emsp;emsp;emsp;<div n="4"></div>')
	text1 = text1.replace('<div n="5"/>', 'emsp;emsp;emsp;emsp;emsp;<div n="5"></div>')
	text1 = re.sub('<div n="([^"]*)"/>', '<div n="\g<1>"></div>', text1)
	text1 = text1.replace('<lb/>', '<br />')
	# AP90 compounds and meanings break
	text1 = text1.replace('<b>--', '<br /><b>--')
	text1 = text1.replace('<span class="s">--', '<br /><span class="s">--')
	# — breaks
	text1 = text1.replace('— ', '<br />— ')
	return text1


def longest_string(text):
	"""Return the longest tring from input regular expression."""
	
	splts = re.split('[.*+]+', text)
	longest = ''
	for splt in splts:
		if len(splt) > len(longest):
			longest = splt
	return longest


def block1(data, inTran='slp1', outTran='slp1'):
	"""Return a dict from input data."""

	root = ET.fromstring(data)
	key1 = root.findall("./h/key1")[0].text
	key2 = root.findall("./h/key2")[0].text
	pc = root.findall("./tail/pc")[0].text
	lnum = root.findall("./tail/L")[0].text
	m = re.split('<body>(.*)</body>', data)
	text = m[1]
	text1 = convert_sanskrit(text, inTran, outTran)
	return {'key1': key1, 'key2': key2, 'pc': pc, 'text': text, 'modifiedtext': text1, 'lnum': lnum}


def block2(dictlist, query, inTran='slp1', outTran='devanagari'):
	"""Return a dict with dictcode as key and a list of block1 as value."""
	
	final = {}
	for dictionary in dictlist:
		sqlitepath = find_sqlite(dictionary)
		# Connect to DB.
		con = sqlite3.connect(sqlitepath)
		# Query.
		query1 = "SELECT * FROM " + dictionary + " WHERE " + query
		# Execute query.
		ans = con.execute(query1)
		# Initialize a blank list.
		result = []
		# For all results,
		for [headword, lnum, data] in ans.fetchall():
			# Append to list.
			result.append(block1(data, inTran, outTran))
		# dictionary code as key and list as value.
		final[dictionary]  = result
	return final


def block3(dictlist, reg, inTran='slp1', outTran='devanagari'):
	"""Return a dict with dictcode as key and a list of block1 as value. Regex matched."""
	
	final = {}
	for dictionary in dictlist:
		sqlitepath = find_sqlite(dictionary)
		con = sqlite3.connect(sqlitepath)
		# Read all entries.
		ans = con.execute("SELECT * FROM " + dictionary)
		# Longest string
		longestString = longest_string(reg)
		result = []
		# For all entries in a dictionary,
		for [headword, lnum, data] in ans.fetchall():
			# If the longestString in headword,
			if longestString in headword:
				# And matches the regex,
				if re.search(reg, headword):
					# Append to result.
					result.append(block1(data, inTran, outTran))
		final[dictionary] = result
	return final


@api.route('/' + apiversion + '/dicts/<string:dictionary>/lnum/<string:lnum>')
@api.doc(params={'dictionary': 'Dictionary code.', 'lnum': 'L number.'})
class DL(Resource):
	"""Return result for given dictionary and lnum."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, dictionary, lnum):
		dictlist = [dictionary]
		final = block2(dictlist, "lnum = " + str(lnum))
		return jsonify(final)
 

@api.route('/' + apiversion + '/dicts/<string:dictionary>/hw/<string:hw>')
@api.doc(params={'dictionary': 'Dictionary code.', 'hw': 'Headword to search.'})
class DH(Resource):
	"""Return result for given dictionary and headword."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, dictionary, hw):
		dictlist = [dictionary]
		final = block2(dictlist, "key = " + "'" + hw + "'")
		return jsonify(final)


@api.route('/' + apiversion + '/dicts/<string:dictionary>/reg/<string:reg>')
@api.doc(params={'dictionary': 'Dictionary code.', 'reg': 'Find the headwords matching the given regex.'})
class DR(Resource):
	"""Return result for given dictionary and regex."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, dictionary, reg):
		dictlist = [dictionary]
		final = block3(dictlist, reg)
		return jsonify(final)


@api.route('/' + apiversion + '/hw/<string:hw>')
@api.doc(params={'hw': 'Headword to search in all dictionaries.'})
class H(Resource):
	"""Return the entries of headword from all dictionaries."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, hw):
		dictlist = cologne_dicts
		final = block2(dictlist,  "key = " + "'" + hw + "'")
		return jsonify(final)


@api.route('/' + apiversion + '/hw/<string:hw>/<string:inTran>/<string:outTran>')
@api.doc(params={'hw': 'Headword to search.', 'inTran': 'Input transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis', 'outTran': 'Output transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class HIO(Resource):
	"""Return the entries of this headword from all dictionaries with specified input and output transliteration."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, hw, inTran, outTran):
		dictlist = cologne_dicts
		final = block2(dictlist, "key = " + "'" + hw + "'", inTran, outTran)
		return jsonify(final)


@api.route('/' + apiversion + '/reg/<string:reg>')
@api.doc(params={'reg': 'Regex to search in all dictionaries.'})
class R(Resource):
	"""Return the entries of this regex from all dictionaries."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, reg):
		dictlist = cologne_dicts
		final = block3(dictlist, reg)
		return jsonify(final)


@api.route('/' + apiversion + '/dicts/<string:dictionary>/hw/<string:hw>/<string:inTran>/<string:outTran>')
@api.doc(params={'dictionary': 'Dictionary code.', 'hw': 'Headword to search.', 'inTran': 'Input transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis', 'outTran': 'Output transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class DHIO(Resource):
	"""Return the entries of this headword from specified dictionary with specified input and output transliteration."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, dictionary, hw, inTran, outTran):
		hw = sanscript.transliterate(hw, inTran, 'slp1')
		dictlist = [dictionary]
		final = block2(dictlist, "key = " + "'" + hw + "'", inTran, outTran)
		return jsonify(final)



@api.route('/' + apiversion + '/reg/<string:reg>/<string:inTran>/<string:outTran>')
@api.doc(params={'reg': 'Regex to search.', 'inTran': 'Input transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis', 'outTran': 'Output transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class RIO(Resource):
	"""Return the entries of this regex from all dictionaries with specified input and output transliteration."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, reg, inTran, outTran):
		reg = sanscript.transliterate(reg, inTran, 'slp1')
		dictlist = cologne_dicts
		final = block3(dictlist, reg, inTran, outTran)
		return jsonify(final)


@api.route('/' + apiversion + '/dicts/<string:dictionary>/reg/<string:reg>/<string:inTran>/<string:outTran>')
@api.doc(params={'dictionary': 'Dictionary code.', 'reg': 'Regex to search.', 'inTran': 'Input transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis', 'outTran': 'Output transliteration. devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class DRIO(Resource):
	"""Return the entries of this regex from specified dictionary with specified input and output transliteration."""

	get_parser = reqparse.RequestParser()

	@api.expect(get_parser, validate=True)
	def get(self, dictionary, reg, inTran, outTran):
		reg = sanscript.transliterate(reg, inTran, 'slp1')
		dictlist = [dictionary]
		final = block3(dictlist, reg, inTran, outTran)
		return jsonify(final)


if __name__ == "__main__":
	app.run(debug=True)
	

File: api_trial.php
<?php
// Exclude WARNING messages also, to solve Peter Scharf Mac version.
error_reporting(E_ALL & ~E_NOTICE & ~E_WARNING);
?>
<?php
require_once('getwordXmlClass.php');
require_once('dbgprint.php');
if (isset($_GET['callback'])) {
 header('content-type: application/json; charset=utf-8');
}
header("Access-Control-Allow-Origin: *");

#echo "<br>url=$url";
#$apiObj = API($url);
function getapiCall($url) {
  $temp = new API($url);
  $ans = $temp->ans;
  $table1 = $ans['final'];
  #print_r($table1);
  if (isset($_GET['callback'])) {
   $json = json_encode($table1);
   echo "{$_GET['callback']}($json)";
  }else {
   echo json_encode($table1);
  }
 }
#$url = $_REQUEST['url'];   # problem with ? (wildcard for search)
$uri = $_SERVER['REQUEST_URI'];  
// e.g. uri=/cologne/api/dicts/mw/reg/rA?an/slp1/devanagari
// delete up through /api/
$url = preg_replace('|^.*?/api/|','',$uri);
#dbgprint(true,"api_trial.php: uri=$uri\n");
#dbgprint(true,"api_trial.php: url=$url\n");
 getapiCall($url);

?>
<?php
/*
api.route('/' + apiversion + '/dicts/<string:dictionary>/lnum/<string:lnum>')
api.route('/' + apiversion + '/dicts/<string:dictionary>/hw/<string:hw>')
api.route('/' + apiversion + '/dicts/<string:dictionary>/reg/<string:reg>')
api.route('/' + apiversion + '/dicts/<string:dictionary>/hw/<string:hw>/<string:inTran>/<string:outTran>')
api.route('/' + apiversion + '/dicts/<string:dictionary>/reg/<string:reg>/<string:inTran>/<string:outTran>')
api.route('/' + apiversion + '/hw/<string:hw>')
api.route('/' + apiversion + '/hw/<string:hw>/<string:inTran>/<string:outTran>')
api.route('/' + apiversion + '/reg/<string:reg>')
api.route('/' + apiversion + '/reg/<string:reg>/<string:inTran>/<string:outTran>')
*/
class API {
 public $version;
 public $ans;
 public function __construct($url)  {
  $this->ans = array(); //associative array 
  $this->ans['status'] = 404; // problem
  $this->ans['url'] = $url;
  $this->ans['final'] = "";
  #$this->ans['html'] = "";
  $flag = $this->parse_url();
  if (! $flag) {return;}
  $this->ans['final'] = $this->block2();
 }
 public function  parse_url() {
  $url_parts = explode("/",$this->ans['url']);
  # first (optional) parameter is version
  if (preg_match('/^v[0-9.]+$/',$url_parts[0])) {
   $this->version = $url_parts[0];
   $url_parts = array_slice($url_parts,1); # remove 1st
  } else {
   $this->version = "v0.0.1"; // latest version?
  }
  $this->ans['version'] = $this->version;
  # dicts
  #print_r($url_parts); echo "<br/>";
  if($url_parts[0] == 'dicts') {
   $url_parts = array_slice($url_parts,1);
   if (count($url_parts) == 0) { // error
    return False;
   }
   $this->ans['dict'] = $url_parts[0];
   $url_parts = array_slice($url_parts,1);
  }else {
  // if first parameter is not 'dicts', then it is implied to be 'all'
   $this->ans['dict'] = 'all';
  }
  // dictlist is an array of dictionary codes
  $cologne_dicts = array('acc', 'ae', 'ap90', 'ben', 'bhs', 'bop', 'bor', 'bur', 'cae', 'ccs', 'gra', 'gst', 'ieg', 'inm', 
    'krm', 'mci', 'md', 'mw', 'mw72', 'mwe', 'pe', 'pgn', 'pui', 'pw', 'pwg', 'sch', 'shs', 'skd', 'snp', 
    'stc', 'vcp', 'vei', 'wil', 'yat');
  if ($this->ans['dict'] == 'all') {
   $this->ans['dictlist'] = $cologne_dicts;
  }else if (! in_array($this->ans['dict'],$cologne_dicts)) {
   # bad dictionary code
   return False;
  }else {
   $this->ans['dictlist'] = array($this->ans['dict']);
  }
  if (count($url_parts) == 0) { // error
    return False;
  }
  // next parameter hw or reg or lnum
  #$this->ans['hw'] = null;
  #$this->ans['reg'] = null;
  #$this->ans['lnum'] = null;
  if (! in_array($url_parts[0],array('hw','reg','lnum'))) {
   return False; // error
  }
  if (count($url_parts) == 1) {
   // error. no value given  for hw, etc.
   return False;
  }
  $this->ans[$url_parts[0]] = $url_parts[1];
  $this->ans['key'] = $this->ans['hw'];  # key is a 'synonym' of 'hw' that is used in getword api
  $this->ans['regex'] = $this->ans['reg']; # regex is used in getword_xml
  $url_parts = array_slice($url_parts,2);
  // The other url parts, if any, are the inTran and outTran parts
  if (count($url_parts) == 0) {
   // set defaults for inTran and outTran. Name them acc. to getword api
   $this->ans['input'] = 'slp1';
   $this->ans['output'] = 'slp1';
  }else if (count($url_parts) != 2) {
   # error condition
   return False;
  }else {
   $inTran = $url_parts[0];
   $outTran = $url_parts[1];
   $input = $inTran;
   $output = $outTran;
   $this->ans['inTran'] = $input;
   $this->ans['outTran'] = $output;
   if ($inTran == 'devanagari') {$input = 'deva';}
   if ($outTran == 'devanagari') {$output = 'deva';}
   $this->ans['input'] = $input;
   $this->ans['output'] = $output;   
  }
  return True;
 }
 public function block2() {
  /* construct $_REQUEST parameters for getword_xml call. 
     Alternately could use curl.
  */
 $possible_parms = ['key','lnum','regex','input','output'];
 foreach($possible_parms as $parm) {
  if (isset($this->ans[$parm])) {
   $value = $this->ans[$parm];
   $_REQUEST[$parm] = $value;
  }
 }
 $dictlist = $this->ans['dictlist'];
 $final = array(); //associative array, dictionary code is key
 $dbg=False;
 foreach($dictlist as $dict) {
  $_REQUEST['dict'] = $dict;
  $temp = new GetwordXmlClass();
  $jsonobj = $temp->json;
  $obj = json_decode($jsonobj,$assoc=True);
  $status = $obj['status'];
  dbgprint($dbg,"api_trial.php: $dict, $status\n");
  $result=array();
  if ($status  == 200 ) {
   #jsonobj = r.json()
   $datarr = $obj['xml'];  # list of data string
   $htmlarr = $obj['html'];
   for ($idata = 0; $idata < count($datarr); $idata++) {
    $data = $datarr[$idata];
    $html = $htmlarr[$idata];
    $result[] = $this->block1($data,$html);
   }
   $final[$dict] = $result;
  } 
 }
 #$final_json = json_encode($final);
 return $final;
 }
 public function block1($data,$html) {
 if (preg_match('|<key1>(.*?)</key1>|',$data,$matches)) {
  $key1 = $matches[1];
 } else {
  $key1 = 'key1 ?';
 }
 if (preg_match('|<key2>(.*?)</key2>|',$data,$matches)) {
  $key2 = $matches[1];
 } else {
  $key2 = 'key2 ?';
 }
 if (preg_match('|<tail>.*?<L.*?>(.*?)</L>|',$data,$matches)) {
  $lnum = $matches[1];
 } else {
  $lnum = 'lnum ?';
 }
 if (preg_match('|<tail>.*?<L.*?>(.*?)</L>|',$data,$matches)) {
  $lnum = $matches[1];
 } else {
  $lnum = 'lnum ?';
 }
 if (preg_match('|<tail>.*?<pc.*?>(.*?)</pc>|',$data,$matches)) {
  $pc = $matches[1];
 } else {
  $pc = 'pc ?';
 }
 if (preg_match('|<body>(.*?)</body>|',$data,$matches)) {
  $body = $matches[1];
 } else {
  $body = 'body ?';
 }
 $text = $body;
 $htmla = preg_replace('|^<info>.*?</info>|','',$html);
 if (preg_match('|<body>(.*?)</body>|',$htmla,$matches)) {
  $text1 = $matches[1];
 } else {
  $text1 = 'html body ?';
 }
 $ans = array('key1'=> $key1, 'key2'=> $key2, 'pc'=> $pc, 'text'=> $text, 
    'modifiedtext'=> $text1, 'lnum'=> $lnum );
 return $ans;
}
}
?>

